angular-ngc/defs.bzl

```diff
@@ -0,0 +1,346 @@
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("@aspect_rules_js//js:defs.bzl", "js_library")
+load("@aspect_rules_js//npm:defs.bzl", "npm_package")
+load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory")
+load("@aspect_rules_esbuild//esbuild:defs.bzl", "esbuild")
+load("@npm//:history-server/package_json.bzl", history_server_bin = "bin")
+load("@npm//:html-insert-assets/package_json.bzl", html_insert_assets_bin = "bin")
+load("@npm//:karma/package_json.bzl", _karma_bin = "bin")
+load("//tools:ng.bzl", "ng_esbuild", "ng_project")
+load("//tools:karma.bzl", "generate_karma_config", "generate_test_bootstrap", "generate_test_setup")
+
+# Common dependencies of Angular applications
+POLYFILLS_DEPS = [
+    "//:node_modules/zone.js",
+]
+
+APPLICATION_DEPS = [
+    "//:node_modules/@angular/common",
+    "//:node_modules/@angular/core",
+    "//:node_modules/@angular/router",
+    "//:node_modules/@angular/platform-browser",
+    "//:node_modules/rxjs",
+    "//:node_modules/tslib",
+] + POLYFILLS_DEPS
+
+APPLICATION_HTML_ASSETS = ["styles.css", "favicon.ico"]
+
+# Common dependencies of Angular packages
+PACKAGE_DEPS = [
+    "//:node_modules/@angular/common",
+    "//:node_modules/@angular/core",
+    "//:node_modules/@angular/router",
+    "//:node_modules/rxjs",
+    "//:node_modules/tslib",
+]
+
+TEST_DEPS = APPLICATION_DEPS + [
+    "//:node_modules/@angular/compiler",
+    "//:node_modules/@types/jasmine",
+    "//:node_modules/jasmine-core",
+    "//:node_modules/@angular/platform-browser-dynamic",
+]
+
+# Common dependencies of Angular test suites using jasmine
+TEST_RUNNER_DEPS = [
+    "//:node_modules/karma-chrome-launcher",
+    "//:node_modules/karma",
+    "//:node_modules/karma-jasmine",
+    "//:node_modules/karma-jasmine-html-reporter",
+    "//:node_modules/karma-coverage",
+]
+
+NG_DEV_DEFINE = {
+    "process.env.NODE_ENV": "'development'",
+    "ngJitMode": "false",
+}
+NG_PROD_DEFINE = {
+    "process.env.NODE_ENV": "'production'",
+    "ngDevMode": "false",
+    "ngJitMode": "false",
+}
+
+def ng_application(name, deps = [], test_deps = [], assets = None, html_assets = None, visibility = ["//visibility:public"], **kwargs):
+    """
+    Bazel macro for compiling an Angular application. Creates {name}, test, serve targets.
+
+    Projects structure:
+      main.ts
+      index.html
+      polyfills.ts (optional)
```

Comment:
doesn't look optional in the macro; targets would be invalid if pollyfills.ts doesn't exist look like

Reply:
I'll just drop the " (optional)" for now to avoid making the macro any more complicated